README
== Overview ==
  Set of programs to convert binary encoded torrent files to human readable
json and back.
  Acts as standart unix "filter" program: reads stdin, writes to stdout.

== Requirements ==
cmake - for building only, can be removed later
yajl - yet another json library (http://lloyd.github.com/yajl/)

== Compiling ==
$ cmake .
$ make
# make install

You may also do "out of tree" builds, like this:
$ mkdir .build && cd .build
$ cmake ..
$ make
# make install
... and all changes: new files, object modules, Makefile's, and so on will
be located in this dir.

== Authors ==
This utilites originally written by two peoples: michail and h4tr3d.
If you want to contact them look there -> http://linuxdv.ru



ChangeLog
0.01:
  First release


orig/torwrite.c
/* $Id: torwrite.c 529 2009-09-11 08:44:53Z michael $ */
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <stdint.h>
#include <stdlib.h>

const uint8_t sp=' ';
const uint8_t ret='\n';
const uint8_t quo='\"';
const uint8_t scr='\\';
const uint8_t hxp='x';
const uint8_t blst='[';
const uint8_t elst=']';
const uint8_t bdct='(';
const uint8_t edct=')';
const uint8_t eqv='=';
const int stdout=1;

enum obj_type {OBJ_NUL, OBJ_INT, OBJ_STR, OBJ_LST, OBJ_DCT};

size_t IntLen(const uint8_t* d, size_t shift, size_t len)
{
 size_t l=1;

 while(d[shift+l]>=48 && d[shift+l]<=57)
 {
  l++;
  if(shift+l==len) exit(2);
 }
 return l;
}

size_t StrLen(const uint8_t* d, size_t shift, size_t len)
{
 size_t l=0;
 size_t sh=shift+1;

 while(1)
 {
  if(d[sh]==quo) break;
  if(d[sh]==scr)
  {
   if(d[sh+1]=='x') sh+=3;
   else sh++;
  }
  sh++;
  l++;
  if(sh>=len) exit(2);
 }
 return l;
}

void PrintDec(size_t n)
{
 size_t num=n,rem;
 size_t len=0;
 uint8_t *out,*pout;

 while(num!=0)
 {
  num/=10;
  len++;
 }

 out=malloc(len);
 pout=out+len-1;
 num=n;

 while(num!=0)
 {
  rem=num%10;
  num/=10;
  *pout--=rem+'0';
 }
 write(stdout,out,len);
 free(out);
}

uint8_t Hex2N(uint8_t h)
{
 if(h>='0' && h<='9') return h-'0';
 if(h>='A' && h<='F') return h-'A'+10;
 if(h>='a' && h<='f') return h-'a'+10;
 exit(2);
}

uint8_t Hex2Sym(const uint8_t *d, size_t sh)
{
 return (Hex2N(d[sh])<<4)|Hex2N(d[sh+1]);
}

int main(int argc, char** argv)
{
 if(argc!=2) return 1;

 int fd;
 struct stat st;
 uint8_t *pdata;
 uint8_t ch,c;
 size_t shift=0,len;

 fd=open(argv[1],O_RDONLY); if(fd==-1) return 1;
 if(fstat(fd,&st)!=0) return 1;
 if(st.st_size==0) return 1;
 pdata=(uint8_t*) mmap(0,st.st_size,PROT_READ,MAP_SHARED,fd,0);

 while(shift<st.st_size)
 {
  c=pdata[shift];
  if(c==bdct) {ch='d'; write(stdout,&ch,1);}
  if(c==blst) {ch='l'; write(stdout,&ch,1);}
  if(c==edct || pdata[shift]==elst) {ch='e'; write(stdout,&ch,1);}
  if((c>=48 && c<=57) || c=='+' || c=='-')
  {
   len=IntLen(pdata,shift,st.st_size);
   ch='i'; write(stdout,&ch,1);
   write(stdout,pdata+shift,len);
   ch='e'; write(stdout,&ch,1);
   shift+=len-1;
  }
  if(c==quo)
  {
   len=StrLen(pdata,shift,st.st_size);
   PrintDec(len); ch=':'; write(stdout,&ch,1);
   shift++;
   while(1)
   {
    c=pdata[shift];
    if(c==quo) break;
    if(c==scr)
    {
     shift++;
     c=pdata[shift];
     if(c=='x')
     {
      shift++;
      ch=Hex2Sym(pdata,shift);
      shift++;
     }
     else
     {
      switch(c)
      {
      case('n'): {ch='\n'; break;}
      case('a'): {ch='\a'; break;}
      case('b'): {ch='\b'; break;}
      case('e'): {ch='\e'; break;}
      case('f'): {ch='\f'; break;}
      case('r'): {ch='\r'; break;}
      case('t'): {ch='\t'; break;}
      case('v'): {ch='\v'; break;}
      default: ch=c;
      }
     }
    }
    else ch=c;
    write(stdout,&ch,1);
    shift++;
   }
  }
  shift++;
 }

 munmap(pdata,st.st_size);
 close(fd);
 return 0;
}


src/json2torrent.c
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include <yajl/yajl_gen.h>
#include <yajl/yajl_parse.h>

/* FD's */
FILE *in;
FILE *out;

struct yajl_handle_t *handle = NULL;
struct yajl_gen_t *gen = NULL;

#define BUF_SIZE (64 * 1024)

static int
write_integer(void *ctx, long long l)
{ return (fprintf(out, "i%llie", l) > 0 ? 1 : 0); }

static int
write_dict_start(void *ctx)
{ return fputc('d', out); }

static int
write_string(void *ctx, const unsigned char *str, size_t len)
{ return (!!fprintf(out, "%u:", len) && !!fwrite(str, 1, len, out)); }

static int
write_list_start(void *ctx)
{ return fputc('l', out); }

/* for dictionary, list and integer *
 * end we have the same marker: 'e' */
static int
write_end(void *ctx)
{ return fputc('e', out); return 0; }

yajl_callbacks callbacks = {
  NULL,             /* null    */
  NULL,             /* bool    */
  write_integer,    /* integer */
  NULL,             /* double  */
  NULL,             /* number  */
  write_string,     /* string  */
  write_dict_start, /* start map   */
  write_string,     /* map key     */
  write_end,        /* end map     */
  write_list_start, /* start array */
  write_end         /* end array   */
};

void
check_yajl_error(struct yajl_handle_t *handle, unsigned char *buf, size_t read)
{
  unsigned char *yajl_error = NULL;

  if ((yajl_error = yajl_get_error(handle, 1, buf, read)) != NULL)
  {
    fwrite(buf, read, 1, stderr);
    fputc('\n', stderr);
    yajl_free_error(handle, yajl_error);
  }
}

int
main(int argc, char **argv)
{
  static unsigned char buf[BUF_SIZE];
  size_t read = 0;
  /* FILE */ in  = stdin;
  /* FILE */ out = stdout;

  gen = yajl_gen_alloc(NULL);
  yajl_gen_config(gen, yajl_gen_validate_utf8, 0);

  handle = yajl_alloc(&callbacks, NULL, (void *) gen);

  yajl_config(handle, yajl_allow_comments,        1);
  yajl_config(handle, yajl_dont_validate_strings, 1);

  while ((read = fread(buf, 1, (BUF_SIZE - 1), in)) != 0)
  {
    buf[read] = '\0';

    if (yajl_parse(handle, buf, read) != yajl_status_ok)
    {
      check_yajl_error(handle, buf, read);
      break;
    }
  }

  if (yajl_complete_parse(handle) != yajl_status_ok)
    check_yajl_error(handle, buf, read);

  yajl_free(handle);

  exit(EXIT_SUCCESS);
}



src/torrent2json.c
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include <yajl/yajl_gen.h>

/* FD's */
FILE *in;
FILE *out;

struct yajl_gen_t *gen = NULL;

void (*select_handler(int))(int);

/* handlers */
void
get_integer(int chr)
{ /* 'chr' always 'i' in this case */
  int c = '\0';
  bool cont = true;
  bool negative = false;
  long int value = 0; /* signed */

  while (cont)
  {
    c = fgetc(in);
    switch ((isdigit(c) != 0) ? '0' : c) /* little hack here */
    {
      case '0' :
        value *= 10; /* (0 * 10) -> 0, so it works correct */
        value += c - '0';
        break;
      case 'e' : /* integer end marker */
        cont = false;
        break;
      case '-' :
        if (value == 0) /* true if '-' - first char after 'i' */
        {
          negative = true;
          break;
        } /* else we consider it as garbage  */
      default  :
        fprintf(stderr, "Garbage after integer: %i%c<\n", value, c);
        exit(EXIT_FAILURE);
        break;
    }
  }

  if (negative) value *= -1;
  yajl_gen_integer(gen, value);
}

void
get_string(int chr)
{
  /* in this case 'chr' first digit of string lenght */
  unsigned int len = (unsigned char) chr - '0';
  int c = '\0';
  unsigned int i = 0;
  unsigned char *buf = NULL;

  /* first, get string lenght */
  while (isdigit((c = fgetc(in))))
    len *= 10, len += c - '0';
  /* 'c' now should contain ':' */

  if ((buf = malloc(sizeof(char) * (len + 1))) == NULL)
  {
    fprintf(stderr, "Can't allocate memory, exiting.\n");
    exit(EXIT_FAILURE);
  }

  memset(buf, '\0', sizeof(char) * (len + 1));
  for (i = 0; i < len; i++)
    buf[i] = fgetc(in);

  yajl_gen_string(gen, buf, len);

  free(buf);
}

void
get_dict(int chr)
{ /* 'chr' always 'd' in this case */
  void (*handler)(int) = NULL;
  int c = '\0';

  yajl_gen_map_open(gen);

  while ((c = fgetc(in)) != 'e')
  {
    handler = select_handler(c);
    if (handler) handler(c);
  }

  yajl_gen_map_close(gen);
}

void
get_list(int chr)
{ /* 'chr' always 'l' in this case */
  void (*handler)(int) = NULL;
  int c = '\0';

  yajl_gen_array_open(gen);

  while ((c = fgetc(in)) != 'e')
  {
    handler = select_handler(c);
    if (handler) handler(c);
  }

  yajl_gen_array_close(gen);
}

/* general purpose functions */
void (*select_handler(int chr))(int)
{
  switch ((isdigit(chr) != 0) ? '0' : chr)
  {
    case 'd' : /* dictionary */
      return get_dict;
      break;
    case 'i' : /* integer */
      return get_integer;
      break;
    case 'l' : /* list */
      return get_list;
      break;
    case '0' : /* string */
      return get_string;
      break;
    default:
      fprintf(stderr, "Unknown marker: %u\n", chr);
      break;
  }

  return NULL;
}

int
main(int argc, char **argv)
{
  void (*handler)(int) = NULL;
  int c = '\0';
  const unsigned char *buf = NULL;
  size_t len = 0;

  /* FILE */ in  = stdin;
  /* FILE */ out = stdout;

  gen = yajl_gen_alloc(NULL);

  yajl_gen_config(gen, yajl_gen_beautify,         1);
  yajl_gen_config(gen, yajl_gen_indent_string, "\t");

  while ((c = fgetc(in)) != EOF)
  {
    handler = select_handler(c);
    if (handler) handler(c);
  }

  yajl_gen_get_buf(gen, &buf, &len);
  if (len > 0 && buf) fprintf(out, "%s", buf);
  yajl_gen_free(gen);

  exit(EXIT_SUCCESS);
}



orig/torread.c
/* $Id: torread.c 529 2009-09-11 08:44:53Z michael $ */
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <stdint.h>
#include <stdlib.h>

const uint8_t sp=' ';
const uint8_t ret='\n';
const uint8_t quo='\"';
const uint8_t scr='\\';
const uint8_t hxp='x';
const uint8_t blst='[';
const uint8_t elst=']';
const uint8_t bdct='(';
const uint8_t edct=')';
const uint8_t eqv='=';
const int stdout=1;

enum obj_type {OBJ_NUL, OBJ_INT, OBJ_STR, OBJ_LST, OBJ_DCT};

void Byte2Hex(uint8_t ch);
void PrintSpc(int num);
enum obj_type DetType(const uint8_t* d, size_t shift);
size_t FindSym(const uint8_t* d, uint8_t s, size_t shift, size_t len);
size_t PrintInt(const uint8_t* d, size_t shift, size_t len);
size_t PrintStr(const uint8_t* d, size_t shift, size_t len);
size_t PrintLst(const uint8_t* d, size_t shift, size_t len, int skip);
size_t PrintDct(const uint8_t* d, size_t shift, size_t len, int skip);
size_t PrintObj(const uint8_t* d, size_t shift, size_t len, int skip);

void Byte2Hex(uint8_t ch)
{
 static const uint8_t cnv[]={'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
 write(stdout,cnv+((ch&(15<<4))>>4),1);
 write(stdout,cnv+(ch&15),1);
}

void PrintSpc(int num)
{
 int i;
 for(i=0;i<num;i++) write(stdout,&sp,1);
}

enum obj_type DetType(const uint8_t* d, size_t shift)
{
// write(2,d+shift,1);
 if(d[shift]=='i') return OBJ_INT;
 if(d[shift]=='l') return OBJ_LST;
 if(d[shift]=='d') return OBJ_DCT;
 if(d[shift]>=49 && d[shift]<=57) return OBJ_STR;
 return OBJ_NUL;
}

size_t FindSym(const uint8_t* d, uint8_t s, size_t shift, size_t len)
{
 size_t i;
 
 for(i=shift;i<len;i++) if(d[i]==s) return i;
 exit(2);
}

size_t PrintObj(const uint8_t* d, size_t shift, size_t len, int skip)
{
 switch(DetType(d,shift))
 {
  case(OBJ_NUL): exit(2);
  case(OBJ_INT): return PrintInt(d,shift,len);
  case(OBJ_STR): return PrintStr(d,shift,len);
  case(OBJ_LST): return PrintLst(d,shift,len,skip);
  case(OBJ_DCT): return PrintDct(d,shift,len,skip);
 }
 exit(2);
}

size_t PrintInt(const uint8_t* d, size_t shift, size_t len)
{
 size_t e=FindSym(d,'e',shift,len);
 write(stdout,d+shift+1,e-shift-1);
 return e+1;
}

size_t PrintStr(const uint8_t* d, size_t shift, size_t len)
{
 size_t e=FindSym(d,':',shift,len);
 int l=atoi((const char*)(d+shift));
 if(e+l>=len) exit(2);
 size_t p;
 
 write(stdout,&quo,1);
 for(p=e+1;p<=e+l;p++)
 {
  if(d[p]==127 || d[p]<32)
  {
   write(stdout,&scr,1);
   write(stdout,&hxp,1);
   Byte2Hex(d[p]);
  }
  else if(d[p]=='\\')
  {
   write(stdout,&scr,1);
   write(stdout,&scr,1);
  }
  else if(d[p]=='\"')
  {
   write(stdout,&scr,1);
   write(stdout,&quo,1);
  }
  else write(stdout,d+p,1);
 }

 write(stdout,&quo,1);
 return e+l+1;
}

size_t PrintLst(const uint8_t* d, size_t shift, size_t len, int skip)
{
 size_t ishift=shift+1;
 
 write(stdout,&blst,1);
 write(stdout,&ret,1);
 
 while(d[ishift]!='e')
 {
  PrintSpc(skip+1);
  ishift=PrintObj(d,ishift,len,skip+1);
  write(stdout,&ret,1);
  if(ishift>=len) exit(2);
 }
 PrintSpc(skip);
 write(stdout,&elst,1);
 return ishift+1;
}

size_t PrintDct(const uint8_t* d, size_t shift, size_t len, int skip)
{
 size_t ishift=shift+1;
 
 write(stdout,&bdct,1);
 write(stdout,&ret,1);
 
 while(d[ishift]!='e')
 {
  PrintSpc(skip+1);
  if(DetType(d,ishift)!=OBJ_STR) exit(2);
  ishift=PrintStr(d,ishift,len);
  write(stdout,&sp,1);
  write(stdout,&eqv,1);
  write(stdout,&sp,1);
  ishift=PrintObj(d,ishift,len,skip+1);
  write(stdout,&ret,1);
  if(ishift>=len) exit(2);
 }
 PrintSpc(skip);
 write(stdout,&edct,1);
 return ishift+1;
}

int main(int argc, char** argv)
{
 if(argc!=2) return 1;

 int fd;
 struct stat st;
 uint8_t *pdata;
 
 fd=open(argv[1],O_RDONLY); if(fd==-1) return 1;
 if(fstat(fd,&st)!=0) return 1;
 if(st.st_size==0) return 1;
 pdata=(uint8_t*) mmap(0,st.st_size,PROT_READ,MAP_SHARED,fd,0);
 PrintDct(pdata,0,st.st_size,0);
 write(stdout,&ret,1);
 munmap(pdata,st.st_size);
 close(fd);

 return 0;
}
